### DISCUSSION TITLE
Cannot override Nginx Configuration in Advanced tab

### DISCUSSION THREAD

**Discussion by martinikolaev** [ZoeyVid/NPMplus#1027] [Q&A] [answered]
I am trying to follow the following guide, but for authelia:
https://www.authelia.com/integration/proxies/nginx-proxy-manager/

Everything is good until i try to overwrite the NGINX Configration in Advanced tab

![image](https://github.com/user-attachments/assets/a6206418-54db-4dec-9bbc-581d10f22f44)

![image](https://github.com/user-attachments/assets/0639cb82-0cb5-495c-8a5a-674434225e50)

Once I save the config:
![image](https://github.com/user-attachments/assets/d9504178-2b96-4f7f-b362-623cf0f0e325)

Status becomes offline.

Logs from the container:
![image](https://github.com/user-attachments/assets/7183c548-9aa5-4bdb-abd1-fd995d67f77e)

Permissions are all okay, I can create new proxies, can modify them and etc, only when I try to add anything in Advanced tab it fails.

Instead when I add stuff to Advanced tab, it deletes the config file (in my case 26.conf) and doesnt add it. When I remove the content from the advanced tab the file is recreated.

For clarity this is the exact text that has been pasted:
`
location / {
    include /snippets/proxy.conf;
    proxy_pass $forward_scheme://$server:$port;
}
`

![image](https://github.com/user-attachments/assets/9a75dff8-e4ff-46eb-881b-5873f0f21694)

`
{
  "id": 27,
  "created_on": "2024-08-19 17:55:43",
  "modified_on": "2024-08-19 17:58:19",
  "owner_user_id": 1,
  "domain_names": [
    "auth.*****"
  ],
  "forward_host": "authelia",
  "forward_port": 9091,
  "access_list_id": 0,
  "certificate_id": 3,
  "ssl_forced": false,
  "caching_enabled": false,
  "block_exploits": false,
  "advanced_config": "location / {\r\n    include /snippets/proxy.conf;\r\n    proxy_pass $forward_scheme://$server:$port;\r\n}",
  "meta": {
    "letsencrypt_agree": false,
    "dns_challenge": false
  },
  "allow_websocket_upgrade": false,
  "http2_support": false,
  "forward_scheme": "http",
  "enabled": 1,
  "locations": [],
  "hsts_enabled": false,
  "hsts_subdomains": false
`

**Comment by Zoey2936** [comment]
Hi, you don't need this advanced config

**Reply by martinikolaev** [reply]
Can you elaborate?

**Reply by Zoey2936** [reply]
the snippet you are including justs sets headers which are alrewady set

**Comment by ne0ark** [comment]
Hello,

I am running into the same issue with Authelia Endpoint.

```location / {
set $upstream_authelia http://SERVERIP:9091; # This example assumes a Docker deployment 
proxy_pass $upstream_authelia;
client_body_buffer_size 128k;

#Timeout if the real server is dead
proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;

# Advanced Proxy Config
send_timeout 5m;
proxy_read_timeout 360;
proxy_send_timeout 360;
proxy_connect_timeout 360;

# Basic Proxy Config
proxy_set_header Host $host;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto $scheme;
proxy_set_header X-Forwarded-Host $http_host;
proxy_set_header X-Forwarded-Uri $request_uri;
proxy_set_header X-Forwarded-Ssl on;
proxy_redirect  http://  $scheme://;
proxy_http_version 1.1;
proxy_set_header Connection "";
proxy_cache_bypass $cookie_session;
proxy_no_cache $cookie_session;
proxy_buffers 64 256k;

# If behind a reverse proxy, forwards the correct IP, assumes you're using Cloudflare. Adjust IP for your Docker network.
set_real_ip_from 172.19.0.0/16;
real_ip_header CF-Connecting-IP;
real_ip_recursive on;
}
```

It doesn't seem like its working. What is the best way to add the endpoint? https://docs.ibracorp.io/authelia/nginx/nginx-config-endpoint

**Reply by Zoey2936** [reply]
Still useless advanced config, can be removed

**Reply by martinikolaev** [reply]
Can you instead provide useful tutorial on how to setup authelia endpoint with NPMPlus?

**Reply by Zoey2936** [reply]
sorry, but it is not my job to fix theier problems

**Reply by Zoey2936** [reply]
I've already recreated a similar config for authentic. And in anycase, the way this authentication works should not be used

**Reply by ne0ark** [reply]
Yes, it seems like the issue was caused by the following:

real_ip_header X-Forwarded-For;
proxy_buffers 64 256k;

Is there any reason NPMplus would have errors when these are used in the custom config?

If you strip these out of the authelia guide, custom config should work. It would be great if NPM did some AI magic to combine the custom config with the existing one.

**Reply by Zoey2936** [reply]
But you don't need any advanced/custom config, just create a new proxy host

**Reply by ne0ark** [reply]
> But you don't need any advanced/custom config, just create a new proxy host

Yes, this works great, but you will still get errors for the endpoint unless you remove the above two config settings.

**Reply by Zoey2936** [reply]
If you leave the advanced config empty then theese lines can't even exist...

**Reply by ne0ark** [reply]
> If you leave the advanced config empty then theese lines can't even exist...

We can't leave it empty for the endpoint. You need to have at least the following for authentication to work:

![image](https://github.com/user-attachments/assets/c6e18199-62ad-4cdb-b031-36b3d1e72d40)

![image](https://github.com/user-attachments/assets/82a85cc2-9afc-400c-98a7-2bb8f7220741)

**Reply by Zoey2936** [reply]
I talk about the config you first posted in this thread

**Reply by ne0ark** [reply]
> I talk about the config you first posted in this thread

Yes, that config was for the endpoint, similar to Authelia-location.conf.

**Reply by DJKatastrof** [reply]
Any progress here? im running into the same error....

**Reply by Zoey2936** [reply]
currently there is only an example for authentic

**Reply by DJKatastrof** [reply]
Thats a bummer.. I was looking forward to change my proxy to this one. Seems like im stuck with regular NPM untill this is solved

**Reply by Zoey2936** [reply]
mainly because their config examples are even worse than the one from authentic

**Reply by DJKatastrof** [reply]
Yea. their documentation is dogshit lol. I agree

**Reply by DJKatastrof** [reply]
I actually solved it. The problem was 

This line. Remove that and it works
       proxy_buffers 64 256k;

**Comment by DJKatastrof** [comment]
If anyone else is interested 
Put this on the proxy.
```
 location / {
        set $upstream_authelia http://CHANGEYOURIP:PORT; 
        proxy_pass $upstream_authelia;
        client_body_buffer_size 128k;
 
        #Timeout if the real server is dead
        proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;
 
        # Advanced Proxy Config
        send_timeout 5m;
        proxy_read_timeout 360;
        proxy_send_timeout 360;
        proxy_connect_timeout 360;
 
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-Host $http_host;
        proxy_set_header X-Forwarded-Uri $request_uri;
        proxy_set_header X-Forwarded-Ssl on;
        proxy_redirect  http://  $scheme://;
        proxy_http_version 1.1;
        proxy_set_header Connection "";
        proxy_cache_bypass $cookie_session;
        proxy_no_cache $cookie_session;

        # If behind reverse proxy, forwards the correct IP
        set_real_ip_from 10.0.0.0/8;
        set_real_ip_from 172.0.0.0/8;
        set_real_ip_from 192.168.0.0/16;
        set_real_ip_from fc00::/7;
        real_ip_header X-Forwarded-For;
        real_ip_recursive on;
    }
```

Put this on your application
```
location /authelia {
    internal;
    set $upstream_authelia http://CHANGEIP:PORT/api/verify; #ADD YOUR IP AND PORT OF AUTHELIA
    proxy_pass_request_body off;
    proxy_pass $upstream_authelia;    
    proxy_set_header Content-Length "";
 
    # Timeout if the real server is dead
    proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;
    client_body_buffer_size 128k;
    proxy_set_header Host $host;
    proxy_set_header X-Original-URL $scheme://$http_host$request_uri;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $remote_addr; 
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host $http_host;
    proxy_set_header X-Forwarded-Uri $request_uri;
    proxy_set_header X-Forwarded-Ssl on;
    proxy_redirect  http://  $scheme://;
    proxy_http_version 1.1;
    proxy_set_header Connection "";
    proxy_cache_bypass $cookie_session;
    proxy_no_cache $cookie_session;

    send_timeout 5m;
    proxy_read_timeout 240;
    proxy_send_timeout 240;
    proxy_connect_timeout 240;
}
 
    location / {
        set $upstream_APPLICATION http://CHANGEIP:PORT;  #CHANGE NAME AND IP AND PORT
        proxy_pass $upstream_APPLICATION;  #change name of the service
 
		auth_request /authelia;
		auth_request_set $target_url $scheme://$http_host$request_uri;
		auth_request_set $user $upstream_http_remote_user;
		auth_request_set $groups $upstream_http_remote_groups;
		proxy_set_header Remote-User $user;
		proxy_set_header Remote-Groups $groups;
		error_page 401 =302 https://auth.CHANGEME.COM/?rd=$target_url; #change YOURDOMAIN.COM to your domain
 
		client_body_buffer_size 128k;
 
		proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;
 
		send_timeout 5m;
		proxy_read_timeout 360;
		proxy_send_timeout 360;
		proxy_connect_timeout 360;
 
		proxy_set_header Host $host;
		proxy_set_header X-Real-IP $remote_addr;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header X-Forwarded-Proto $scheme;
		proxy_set_header X-Forwarded-Host $http_host;
		proxy_set_header X-Forwarded-Uri $request_uri;
		proxy_set_header X-Forwarded-Ssl on;
		proxy_set_header Upgrade $http_upgrade;
		proxy_set_header Connection "upgrade";
		proxy_redirect  http://  $scheme://;
		proxy_http_version 1.1;
		proxy_set_header Connection "";
		proxy_cache_bypass $cookie_session;
		proxy_no_cache $cookie_session;
 
		set_real_ip_from 192.168.1.0/16;
		set_real_ip_from 192.168.67.0/16;
		real_ip_header X-Forwarded-For;
		real_ip_recursive on;

    }
```

**Reply by Zoey2936** [reply]
I edited your post to fix the formatting, but there are some parts in this thing which are either handled by NPMplus or not recommended from me, so is it fine if I edit it and put it in the readme?

**Reply by DJKatastrof** [reply]
Thank you! I struggle with formatting, sorry!
Yes, please do :) I can test them as well

**Reply by DJKatastrof** [reply]
I just noticed what proxy_buffers does. Maybe this is something that's already handled by NPMPlus?

Syntax: proxy_buffers number size;
Default: proxy_buffers 8 4k|8k;
Context: http, server, location

Sets the number and size of the buffers used for reading a response from the proxied server, for a single connection. By default, the buffer size is equal to one memory page. This is either 4K or 8K, depending on a platform.

**Reply by Zoey2936** [reply]
yes this is set by NPMplus

**Reply by DJKatastrof** [reply]
Cool, thanks for the help!

**Reply by Zoey2936** [reply]
so can you try the following:
for authentic itself, configure nothing in advanced conf, keep it just empty
and for the host you want to protect put the following in the advanced config, make sure to adjust the marked parts if needed:
```
location / {
    # if you want to enable websockets
    #proxy_set_header Upgrade $http_upgrade;
    #proxy_set_header Connection $connection_upgrade;

    include conf.d/include/proxy-headers.conf;
    proxy_pass $forward_scheme://$server:$port$request_uri;

    auth_request /internal/authelia/authz;
    auth_request_set $user $upstream_http_remote_user;
    auth_request_set $groups $upstream_http_remote_groups;
    auth_request_set $name $upstream_http_remote_name;
    auth_request_set $email $upstream_http_remote_email;
    proxy_set_header Remote-User $user;
    proxy_set_header Remote-Groups $groups;
    proxy_set_header Remote-Email $email;
    proxy_set_header Remote-Name $name;

    auth_request_set $redirection_url $upstream_http_location;
    error_page 401 =302 $redirection_url;
}

location /internal/authelia/authz {
    include conf.d/include/proxy-headers.conf;
    proxy_pass http://<ip>:<port>/api/verify; # change to ip and port of authelia

    internal;
    proxy_method GET;
    proxy_pass_request_body off;
    proxy_set_header Content-Length "";
    proxy_set_header X-Original-URL $scheme://$host$request_uri;
}
```

**Reply by Zoey2936** [reply]
I've updated my comment, this should now be the everything needed to work, if something does not work, please report back. Thanks!

**Reply by DJKatastrof** [reply]
Hey, the proxy host goes offline when adding this. 

If you comment this line then it goes online, but not accessible
"   proxy_pass $forward_scheme://$server:$port$request_uri; "

**Reply by Zoey2936** [reply]
Can you update to the latest release from yesterday please?

**Reply by DJKatastrof** [reply]
Internally its working.
Externally im getting 302 Found

**Reply by Zoey2936** [reply]
What do you mean with internally abd externally?

**Reply by DJKatastrof** [reply]
Inside my network. When I go to food.domain.com it's working. 
When I use my work computer to go to food.domain.com im getting "302 Found".

**Reply by DJKatastrof** [reply]
Edit: No, its still 302. Switching to my config it works.

Wait. I think Its my fault. Wrong certificate on the host.

**Reply by Zoey2936** [reply]
also when using the "Modern" configuration: https://www.authelia.com/integration/proxies/nginx/#implementation?
so this example:
```
session:
  cookies:
    - domain: 'example.com'
      authelia_url: 'https://auth.example.com'
      default_redirection_url: 'https://www.example.com'
```
not this:
```
default_redirection_url: 'https://www.example.com'
session:
  domain: 'example.com'
```

**Reply by Zoey2936** [reply, marked as answer]
if you have time can you please update to the develop tag of npmplus and try again with the exaple you now can find in the readme? I've added a line for the legacy variation

**Reply by DJKatastrof** [reply]
Hey, how do I just the develop tag?

**Reply by Zoey2936** [reply]
in your compose yaml change zoeyvid/npmplus to zoeyvid/npmplus:develop

**Reply by DJKatastrof** [reply]
My container went into a restarting loop..

**Reply by Zoey2936** [reply]
I see can you pull the develop tag in 6 minutes again and try again please, sorry for the issue

**Reply by DJKatastrof** [reply]
Hey man, im sorry but this took too much effort to get working with authelia and crowdsec. I moved on to caddy sorry.

**Reply by Zoey2936** [reply]
No problem, caddy is also a very good choice

**Comment by DJKatastrof** [comment]
```
services:
  npmplus:
    container_name: npmplus
    image: docker.io/zoeyvid/npmplus:develop
    restart: always
    network_mode: host
```
